istioctl/cmd/proxystatus_v2.go

```diff
@@ -0,0 +1,151 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"text/tabwriter"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/spf13/cobra"
+
+	"istio.io/istio/istioctl/pkg/clioptions"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/adsc"
+)
+
+const (
+	connectionsTypeURL = "istio.io/connections"
+)
+
+func newProxyStatusCommand() *cobra.Command {
+	var opts clioptions.CentralControlPlaneOptions
+
+	statusCmd := &cobra.Command{
+		Use:   "proxy-status [<pod-name[.namespace]>]",
+		Short: "Retrieves the synchronization status of each Envoy in the mesh",
+		Long: `
+Retrieves last sent and last acknowledged xDS sync from Pilot to each Envoy in the mesh
+
+`,
+		Example: `# Retrieve sync status for all Envoys in a mesh
+	istioctl proxy-status
+
+# Retrieve sync diff for a single Envoy and Pilot
+	istioctl proxy-status productpage-v1-679f4bcbbb-6j4j6.default
+`,
+		Aliases: []string{"ps"},
+		RunE: func(c *cobra.Command, args []string) error {
+			if opts.Xds == "" {
+				return fmt.Errorf("--endpoint default not implemented; MUST be supplied; try localhost:15012")
+			}
+			adscConn, err := adsc.Dial(opts.Xds, "", &adsc.Config{
```

Comment:
For XDS address - we need 3 flags:
-xdsAddress - a URL. If it is a https:// URL - just use it as is. 
-certDir - I think istioctl should be able to get a cert for a VM ( few people are working on this). User should use that command to grab an Istio certificate before running this command, and use them in the istioctl. We can use a default ( ~/.istio or ./etc/certs ). That's because Istiod will start requiring authentication.
-xdsLabel or something like this to use port forwarding instead of direct connection - for cases where Istiod is not exposed in the gateway ( i.e. mesh expansion / multicluster are disabled ).

Reply:
For the second - adsc options include this, so just need to pass it.

Reply:
@costinm I was able to get `--cert-dir` working.  I set up a port-forward to 10512 for testing.

My first attempt failed with `transport: authentication handshake failed: x509: certificate is valid for istiod.istio-system.svc, istiod-remote.istio-system.svc, istio-pilot.istio-system.svc, not istio-pilot.istio-system`.  To get this to work I needed to change a line in _adsc.go_.  Please address my TODO on the revised PR.

Reply:
@therealmitchconnors For `--xdsLabel` I believe @costinm is suggesting that istioctl port-forward to any pod chosen by `kubectl -n istio-system get pods --selector istio=pilot`, where --xdsLabel sets the selector.  Is that correct or should we use something like `--revision <rev>` and look up pods using `istio=pilot,istio.io/rev=<revision>`

Reply:
@costinm if we support --xdsLabel that doesn't help us understand if we want Istiod's 15010 (plaintext) or 10512 (tls) port.  Does the port on the matching istiod deserve it's own flag?

Reply:
As usual, there is 'demo' and 'prod'. I'm not particularly concerned with demo - plain text and port forward are all fine. 

For 'prod' - the plaintext ports should not even be enabled ( 15010, 8080 ). There is only 15012. The certificate should be checked - and it should be istiod[-REV].istio-system.svc or the custom address (in which case - standard ACME cert verification).

The next step in the 'event' feed is for Istiod itself to select all the other istiod pods and connect to them - so regardless of which Istiod replica you hit, you should see all the events across all revisions and instances. It must be done on the server side for several reasons:

- there is no guarantee that unpriviledged clients ( app admins ) would have any permission into istio-system, so they won't be able to query by pod or connect at will to any pod.
- istiod itself will use the cross-XDS connections to sync - for example workload entries, etc
- istiod may also use the cross-XDS connections to implement the caching aspects of the new XDS federation we're working on. 

For your use cases - the first is the most important. I think it is a critical requirement for 'app admins' to NOT have any permissions in istio-system ( in any prod env ).

Reply:
@costinm @howardjohn I want to make sure I am doing this correctly.  I understand --xdsAddress and --certDir.  For --xdsLabel am I connecting to a workload sidecar?  An admin pod?  The [Central Istiod doc](https://docs.google.com/document/d/1r58al52otp1uvB7YFHh8ifA_78hYeyd0hNLq4qPDSUM/edit#) does not include _istioctl_ on the block diagrams.

Reply:
you don't need a sidecar - you are connecting to istiod.istio-system.svc, using TLS and certDir. 

If you want to get the XDS config generated by Istiod for a specific workload - you need to pass the labels of the workload, so Istiod generates the correct config.  Istioctl will behave 
like an envoy - that's what ADSC is intended for, emulate the calls envoy makes.

Yes, absolutely Istioctl should have code to get a cert - I believed it already does ( for VM provisioning ). 

The flow of Istioctl is no different than using a mesh expansion VM. 

The makefile we use for VMs ( used to be in samples, now in install tools ) is IMO the best option for now.